### PR TITLE
Remove document.write and use CDTS new declarative setup function

### PR DESCRIPTION
**Pull request by ahmad-shahid**

### CODE REVIEW

File: Components/Entities/IFooterSection.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace GoC.WebTemplate.Components.Entities
 {
-    interface IFooterSection
+    public interface IFooterSection
```
Comment (plup2) [inline Components/Entities/IFooterSection.cs:3]: Not saying it is wrong, just wondering: Why visibility of this interface was changed?

File: WebForms/GoC.WebTemplate/GoCWebTemplate.BilingualErrorPage.Master [hunk 2]
```diff
@@ -12,53 +12,31 @@
         <title><%= WebTemplateModel.Render.HeaderTitle() %></title>
         <meta content="width=device-width,initial-scale=1" name="viewport"/>
         <!-- Load closure template scripts -->
-        <script type="text/javascript" src="<%= string.Concat(WebTemplateModel.CDNPath, "soyutils.js")  %>"></script>
-        <script type="text/javascript" src="<%= string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}wet-{1}.js", WebTemplateModel.CDNPath, WebTemplateModel.TwoLetterCultureLanguage)  %>"></script>
+        <link rel="stylesheet" href="<%= WebTemplateModel.CSSPath%>"/>
+        <script type="text/javascript" src="<%= string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}compiled/wet-{1}.js", WebTemplateModel.CDNPath, WebTemplateModel.TwoLetterCultureLanguage)  %>"data-cdts-setup='<%= WebTemplateModel.Render.ServerSetup() %>'></script>
```
Comment (plup2) [inline WebForms/GoC.WebTemplate/GoCWebTemplate.BilingualErrorPage.Master:16]: Space missign between attributes  (almost didn't catch that one :))

**Review by plup2 — CHANGES_REQUESTED** [review]

File: Components/Entities/IFooterSection.cs (hunk 1 above)
Comment (ahmad-shahid) [reply Components/Entities/IFooterSection.cs:3]: As discussed on call

File: Components.Tests/RenderTests/RenderRefFooterTests.cs [hunk 3]
```diff
@@ -18,8 +18,8 @@ public void RenderRefFooterWithoutSecureSite(Model sut)
         {
             sut.Settings.LeavingSecureSiteWarning.Enabled = false;
 
-            var result = sut.Render.RefFooter(isApplication: false);
-            result.ToString().Should().Be("{\"cdnEnv\":\"\"}");
+            var result = sut.Render.Setup();
+            result.ToString().Should().NotContain("{\"exitScript\":\"\"}");
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/RenderTests/RenderRefFooterTests.cs]: Yes it was too specific. I updated it.

File: Components.Tests/RenderTests/RenderSetupTests.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        /// <summary>
+        /// Tests that apostrophes (') are escaped
+        /// </summary>
+        /// <param name="sut"></param>
+        [Theory, AutoNSubstituteData]
+        public void TestRenderSetupApos(Model sut)
+        {
+            sut.ScreenIdentifier = "This is an 'identifier'";
+            sut.Render.Setup().ToString().Should().Contain("\"screenIdentifier\":\"This is an 'identifier'\"");
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/RenderTests/RenderSetupTests.cs]: Fixed! I updated the Json Serializer settings.

File: Components.Tests/UtilTests/CSSPathTest.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            result.EndsWith("/cdts-styles.css");
+        }
+
+        /// <summary>
+        /// BuildSplashCSSPath
+        /// </summary>
+        [Theory, AutoNSubstituteData]
+        public void TestSplashGCWeb(Model sut)
+        {
+            sut.CdtsEnvironment.Theme = "gcintranet";
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/UtilTests/CSSPathTest.cs]: done

File: Components/Utils/ModelBuilder.cs [hunk 6]
```diff
@@ -18,6 +18,283 @@ public ModelBuilder(IModel model)
...
+            //For v4.0.26.x we have to render this section differently depending on the theme, 
+            //GCIntranet theme renders AppName and AppUrl seperately in GCWeb we render it as a List of Links. 
+            if (_model.CdtsEnvironment.ThemeIsGCWeb())
+            {
+                return new AppTop
+                {
+                    AppName = new List<Link> { _model.ApplicationTitle },
+                    SignIn = _model.Builder.BuildHideableHrefOnlyLink(_model.Settings.SignInLinkUrl, _model.ShowSignInLink),
+                    SignOut = _model.Builder.BuildHideableHrefOnlyLink(_model.Settings.SignOutLinkUrl, _model.ShowSignOutLink),
+                    CdnEnv = _model.CdtsEnvironment.CDN,
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components/Utils/ModelBuilder.cs]: done

File: Components.Tests/UtilTests/CSSPathTest.cs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            var result = modelBuilder.BuildAppCSSPath();
+
+            result.EndsWith("/cdts-esdc-styles.css");
+        }
+
+        [Theory, AutoNSubstituteData]
+        public void TestAppGCIntranetECCC(Model sut)
+        {
+            sut.CdtsEnvironment.Theme = "gcintranet";
+            sut.CdtsEnvironment.SubTheme = "eccc";
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/UtilTests/CSSPathTest.cs:123]: Updated existing tests

File: Components.Tests/UtilTests/CSSPathTest.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            result.EndsWith("/cdts-styles.css");
+        }
+
+        /// <summary>
+        /// BuildAppCSSPath
+        /// </summary>
+        [Theory, AutoNSubstituteData]
+        public void TestAppGCWeb(Model sut)
+        {
+            sut.CdtsEnvironment.Theme = "gcintranet";
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/UtilTests/CSSPathTest.cs]: Fixed

File: WebForms/GoC.WebTemplate/GoCWebTemplate.Application.LeftMenu.Master [hunk 9]
```diff
@@ -12,75 +12,44 @@
         <title><%= WebTemplateModel.Render.HeaderTitle() %></title>
         <meta content="width=device-width,initial-scale=1" name="viewport"/>
         <!-- Load closure template scripts -->
-        <script type="text/javascript" src="<%= string.Concat(WebTemplateModel.CDNPath, "soyutils.js")  %>"></script> 
-        <script type="text/javascript" src="<%= string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}wet-{1}.js", WebTemplateModel.CDNPath, WebTemplateModel.TwoLetterCultureLanguage)  %>"></script> 
+        <link rel="stylesheet" href="<%= WebTemplateModel.AppCSSPath%>"/>
+        <script type="text/javascript" src="<%= string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}compiled/wet-{1}.js", WebTemplateModel.CDNPath, WebTemplateModel.TwoLetterCultureLanguage)  %>" data-cdts-setup='<%= WebTemplateModel.Render.AppSetup() %>'></script> 
         <noscript>
             <!-- Write closure fall-back static file -->
+            <a href="GoCWebTemplate.LeftMenu.Master">GoCWebTemplate.LeftMenu.Master</a>
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply WebForms/GoC.WebTemplate/GoCWebTemplate.Application.LeftMenu.Master]: Not sure why that was added :) I removed it

File: WebForms/GoC.WebTemplate/GoCWebTemplate.Application.Master [hunk 10]
```diff
@@ -12,61 +12,35 @@
...
                 <!-- Write closure fall-back static file -->
                 <%= WebTemplateModel.LoadStaticFile(string.Concat("preFooter-", WebTemplateModel.TwoLetterCultureLanguage, ".html")) %>
             </div>
-            <!-- Write closure template -->
-            <script type="text/javascript">
-                var defPreFooter = document.getElementById("def-preFooter");
-                defPreFooter.outerHTML = wet.builder.preFooter(<%= WebTemplateModel.Render.PreFooter() %>);
-            </script>
         </main>
         <div id="def-footer">
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply WebForms/GoC.WebTemplate/GoCWebTemplate.Application.Master]: done

File: WebForms/GoC.WebTemplate/GoCWebTemplate.Transactional.LeftMenu.Master [hunk 11]
```diff
@@ -12,24 +12,19 @@
...
         <!-- GoC.WebTemplate.Core Version <%= WebTemplateModel.AssemblyVersion %> -->
     </head>
     <body vocab="http://schema.org/" typeof="WebPage">
 	     <%= WebTemplateModel.Render.SessionTimeoutControl() %>
-
-        <div id="def-top">
+        <div id="cdts-def-top">
 		    <!-- Write closure fall-back static file -->
             <%= WebTemplateModel.LoadStaticFile(string.Concat("transactTop-", WebTemplateModel.TwoLetterCultureLanguage, ".html")) %>   
         </div>
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply WebForms/GoC.WebTemplate/GoCWebTemplate.Transactional.LeftMenu.Master:30]: done

File: WebForms/TestWebTemplateLeftMenuPage.aspx.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+using System.Web.UI;
+using System.Web.UI.WebControls;
+
+namespace GoC.WebTemplate.WebForms
+{
+    public partial class TestWebTemplateLeftMenuPage : BasePage
+    {
+        protected void Page_Load(object sender, EventArgs e)
+        {
+
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply WebForms/TestWebTemplateLeftMenuPage.aspx.cs:14]: As discussed, I'll open an issue to fix all of the test pages

File: Components/Entities/IFooterSection.cs (hunk 1 above)
Comment (ahmad-shahid) [reply Components/Entities/IFooterSection.cs:3]: As discussed, to keep visibility in sync with other functions

File: Components.Tests/RenderTests/RenderSetupTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (plup2) [reply Components.Tests/RenderTests/RenderSetupTests.cs]: Cool.  This turns it on for ALL JSON serialization but should be ok (it is the case on Java side after all)

Another manual test to perform not that this is sorted out (if not tested already): Test that any HtmlBodyElements get inserted corrected on page once CDTS/WET has finished initializing.   Let me know here the results

[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Components.Tests/RenderTests/RenderSetupTests.cs]: Yup, just tested it out and it works as expected. The elements gets added at the end of the body.

**Review by plup2 — APPROVED** [review]

**Pull request merged**